coderd/cryptokeys/dbkeycache.go

```diff
@@ -0,0 +1,185 @@
+func WithDBCacheClock(clock quartz.Clock) DBCacheOption {
+	return func(d *DBCache) {
+		d.clock = clock
+	}
+}
+
+// NewDBCache creates a new DBCache. It starts a background
+// process that periodically refreshes the cache. The context should
+// be canceled to stop the background process.
+func NewDBCache(ctx context.Context, logger slog.Logger, db database.Store, feature database.CryptoKeyFeature, opts ...func(*DBCache)) (*DBCache, error) {
+	d := &DBCache{
+		db:      db,
+		feature: feature,
+		clock:   quartz.NewReal(),
+		logger:  logger,
+	}
+	for _, opt := range opts {
+		opt(d)
+	}
+
+	cache, latest, err := d.newCache(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("new cache: %w", err)
+	}
+	d.cache, d.latestKey = cache, latest
+
+	go d.refresh(ctx)
+	return d, nil
+}
+
+// Version returns the CryptoKey with the given sequence number, provided that
+// it is neither deleted nor has breached its deletion date.
+func (d *DBCache) Version(ctx context.Context, sequence int32) (codersdk.CryptoKey, error) {
+	now := d.clock.Now().UTC()
+	d.cacheMu.RLock()
+	key, ok := d.cache[sequence]
+	d.cacheMu.RUnlock()
+	if ok {
+		return checkKey(key, now)
+	}
+
+	d.cacheMu.Lock()
+	defer d.cacheMu.Unlock()
+
+	key, ok = d.cache[sequence]
+	if ok {
+		return checkKey(key, now)
+	}
+
+	key, err := d.db.GetCryptoKeyByFeatureAndSequence(ctx, database.GetCryptoKeyByFeatureAndSequenceParams{
+		Feature:  d.feature,
+		Sequence: sequence,
+	})
+	if xerrors.Is(err, sql.ErrNoRows) {
+		return codersdk.CryptoKey{}, ErrKeyNotFound
+	}
+	if err != nil {
+		return codersdk.CryptoKey{}, err
+	}
+
+	if !key.CanVerify(now) {
+		return codersdk.CryptoKey{}, ErrKeyInvalid
+	}
+
+	// If this key is valid for signing then mark it as the latest key.
+	if key.CanSign(now) && key.Sequence > d.latestKey.Sequence {
+		d.latestKey = key
+	}
+
+	d.cache[sequence] = key
+
+	return db2sdk.CryptoKey(key), nil
+}
+
+// Latest returns the latest valid key for signing. A valid key is one that is
+// both past its start time and before its deletion time.
+func (d *DBCache) Latest(ctx context.Context) (codersdk.CryptoKey, error) {
+	d.cacheMu.RLock()
+	latest := d.latestKey
+	d.cacheMu.RUnlock()
+
+	now := d.clock.Now().UTC()
+	if latest.CanSign(now) {
+		return checkKey(latest, now)
+	}
+
+	d.cacheMu.Lock()
+	defer d.cacheMu.Unlock()
+
+	if latest.CanSign(now) {
+		return checkKey(latest, now)
+	}
+
+	// Refetch all keys for this feature so we can find the latest valid key.
+	cache, latest, err := d.newCache(ctx)
+	if err != nil {
+		return codersdk.CryptoKey{}, xerrors.Errorf("new cache: %w", err)
+	}
+
+	if len(cache) == 0 {
+		return codersdk.CryptoKey{}, ErrKeyNotFound
+	}
+
+	if !latest.CanSign(now) {
+		return codersdk.CryptoKey{}, ErrKeyInvalid
+	}
+
+	d.cache, d.latestKey = cache, latest
+
+	return checkKey(latest, now)
+}
+
+func (d *DBCache) refresh(ctx context.Context) {
+	d.clock.TickerFunc(ctx, time.Minute*10, func() error {
+		cache, latest, err := d.newCache(ctx)
```

Comment:
I can't help but wonder whether it might be better / simpler to use an `AfterFunc` timer to invalidate the cache with a 10 minute timeout.  If we do a fetch at 9m55s, it doesn't make a lot of sense to do another fetch 5s later. (Whenever we fetch, reset the timer to 10m.)

Also, we can just leave the cache empty on startup and on invalidation: the next call to grab a key will populate it.

Reply:
Agreed

Reply:
I think this matters less for the DB implementation but I think for the WSProxy one (which I'll be putting up once this gets merged) we would want to avoid only invalidating the cache since it's guarantees you'll always have a slow request